v0.3.3 -> "sig" parameter not in measurements

Hey guys,

I have been trying out the new release v0.3.3, thanks again for the effort you guys put into this!

I did notice a bug when running with `image.VECTORIZED = True`. It seems to stem from the change made in this commit:
https://github.com/transientskp/pyse/commit/7529c36037a4854ece718b8edf34abcf161a9f16
Here you change the sig parameter from `__setattr__` to `__setitem__`.

Since now the `__setitem__` is triggered, we arrive here: https://github.com/transientskp/pyse/blob/b662f48cd116317de0469581eaea405bc80ef6ac/sourcefinder/extract.py#L283 and find that 'sig' is not in `self.measurements` and we get the following error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/timo/software/astron/pyse/sourcefinder/image.py", line 35, in timed
    result = method(*args, **kw)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/timo/software/astron/pyse/sourcefinder/image.py", line 541, in extract
    return self._pyse(
           ^^^^^^^^^^^
  File "/home/timo/software/astron/pyse/sourcefinder/image.py", line 1323, in _pyse
    param["sig"] = maxis[count] / self.rmsmap.data[tuple(maxposs[count])]
    ~~~~~^^^^^^^
  File "/home/timo/software/astron/pyse/sourcefinder/extract.py", line 289, in __setitem__
    raise AttributeError("Invalid parameter")
AttributeError: Invalid parameter

```
To reproduce this, you could run the following from the root of the repository:

```python
from sourcefinder.accessors import open as pyse_open
from sourcefinder.accessors import sourcefinder_image_from_accessor
from sourcefinder import image
image.VECTORIZED = True
path = "test/data/NCP_sample_image_1.fits"
fits_im = pyse_open(path)
pyse_im = sourcefinder_image_from_accessor(fits_im)
pyse_im.extract(8,8)
```

I imagine the fix would either be to add 'sig' to self.measurements or to revert the change.

Either way, I wish you guys a merry christmas and a happy new year!

Cheers,
Timo

> "can you check?" -> yes it also fails for me on `param.sig` if I run that hatch command.

After some contemplation, I think that **mypy** is wrong in this case, it somehow does not see that we are dealing with a [extract.ParamSet](https://github.com/transientskp/pyse/blob/8bd77c9ae181ade28a0f78ceeed02fe25e90ccd6/sourcefinder/extract.py#L228) instantiation i.e. a [collections.abc.MutableMapping.](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableMapping). 

So that line [1285](https://github.com/transientskp/pyse/blob/8bd77c9ae181ade28a0f78ceeed02fe25e90ccd6/sourcefinder/image.py#L1285C17-L1285C86):
```
                param["sig"] = maxis[count] / self.rmsmap.data[tuple(maxposs[count])]
```
Should in fact be
```
                param.sig = maxis[count] / self.rmsmap.data[tuple(maxposs[count])]
```
which is the current way that vectorized code should mimick its non-vectorized counterpart in [line 224 from the extract module](https://github.com/transientskp/pyse/blob/8bd77c9ae181ade28a0f78ceeed02fe25e90ccd6/sourcefinder/extract.py#L224):
```
        measurement.sig = self.sig()
```
with _measurement_ being a _extract.ParamSet_ instance. 
Oddly enough, **mypy** does not complain about this line.

@suvayu Do you agree with this line of reasoning and do you think we can slightly alter 
`hatch run lint:mypy --install-types --non-interactive --pretty` as part of our CI such that it will not raise
```
sourcefinder/image.py:1285: error: "ParamSet" has no attribute "sig"  [attr-defined]
```
?

--- a/sourcefinder/extract.py
+++ b/sourcefinder/extract.py
@@ -275,6 +275,7 @@ class ParamSet(MutableMapping):
         # More metadata about the fit: only valid for Gaussian fits:
         self.chisq = None
         self.reduced_chisq = None
+        self.sig = None
 
     def __getitem__(self, item):
         return self.measurements[item]
@@ -546,10 +547,6 @@ class ParamSet(MutableMapping):
         # replaced by noise**2 since the threshold should not affect
         # the error from the (corrected) maximum pixel method,
         # while it is part of the expression for rho_sq above.
-        # errorpeaksq = ((frac_flux_cal_error * peak) ** 2 +
-        #                clean_bias_error ** 2 + noise ** 2 +
-        #                utils.maximum_pixel_method_variance(
-        #                    beam[0], beam[1], beam[2]) * peak ** 2)
         errorpeaksq = ((frac_flux_cal_error * peak) ** 2 +
                        clean_bias_error ** 2 + noise ** 2 +
                        max_pix_variance_factor * peak ** 2)
@@ -828,7 +825,7 @@ def source_profile_and_errors(data, threshold, noise,
 
     # Calculate residuals
     # NB this works even if Gaussian fitting fails, we generate the model from
-    # the moments-fit parameters.
+    # the moments parameters.
     gauss_arg = (param["peak"].value,
                  param["xbar"].value,
                  param["ybar"].value,
@@ -1718,9 +1715,10 @@ def source_measurements_pixels_and_celestial_vectorised(num_islands, npixs,
     # array. In this way Numba can infer the shape of the output array.
     fitting.moments_enhanced(sources, noises, chunk_positions, xpositions,
                              ypositions, minimum_widths, npixs, thresholds,
-                             local_noise_levels, maxis, fudge_max_pix_factor,
-                             max_pix_variance_factor, np.array(beam),
-                             beamsize, np.array(correlation_lengths), 0, 0,
+                             local_noise_levels, maxposs, maxis,
+                             fudge_max_pix_factor, max_pix_variance_factor,
+                             np.array(beam), beamsize,
+                             np.array(correlation_lengths), 0, 0,
                              Gaussian_islands, Gaussian_residuals, dummy,
                              moments_of_sources, sig, chisq, reduced_chisq)
 

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -14,6 +14,7 @@ from sourcefinder.stats import newton_raphson_root_finder
 
 FIT_PARAMS = ('peak', 'xbar', 'ybar', 'semimajor', 'semiminor', 'theta')
 
+
 @njit
 def find_true_peak(peak, T, epsilon, msq, maxpix):
     """
@@ -149,9 +150,9 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
 
                 peak = newton_raphson_root_finder(find_true_peak, basevalue,
                                                   low_bound, upp_bound,
+                                                  1e-8, 100,
                                                   threshold, epsilon,
-                                                  semiminor_tmp, basevalue,
-                                                  tol=1e-8, max_iter=100)
+                                                  semiminor_tmp, basevalue)
                 # The corrections below for the semi-major and semi-minor axes are
                 # to compensate for the underestimate of these quantities
                 # due to the cutoff at the threshold.

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -10,8 +10,18 @@ from .gaussian import gaussian, jac_gaussian
 from .stats import indep_pixels
 from sourcefinder.deconv import deconv
 from numba import guvectorize, float64, float32, int32
+from sourcefinder.stats import newton_raphson_root_finder
 
 FIT_PARAMS = ('peak', 'xbar', 'ybar', 'semimajor', 'semiminor', 'theta')
+
+
+def find_true_peak(peak, T, epsilon, msq, maxpix):
+    """
+    This function represents equation 2.67 from Spreeuw's thesis.
+
+    """
+    return np.log(peak / maxpix) - (epsilon / msq) * (1 + np.log(T / peak) /
+                                                      (peak / T - 1))
 
 
 def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
@@ -47,13 +57,6 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
     exceptions.ValueError
         If input contains NaN values.
     """
-    # Are we fitting a -ve or +ve Gaussian?
-    if data.mean() >= 0:
-        # The peak is always underestimated when you take the highest pixel.
-        peak = data.max() * fudge_max_pix_factor
-    else:
-        peak = data.min()
-    ratio = threshold / peak
     total = data.sum()
     x, y = np.indices(data.shape)
     xbar = float((x * data).sum() / total)
@@ -65,54 +68,114 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
     working1 = (xxbar + yybar) / 2.0
     working2 = math.sqrt(((xxbar - yybar) / 2) ** 2 + xybar ** 2)
 
+    if (np.isnan(xbar) or np.isnan(ybar) or np.isnan(working1)
+            or np.isnan(working2)):
+        raise ValueError("Unable to estimate Gauss shape")
+
+    maxpos = np.unravel_index(np.abs(data).argmax(), data.shape)
+
+    # Are we fitting a -ve or +ve Gaussian?
+    if data.mean() >= 0:
+        peak = data[maxpos]
+        # The peak is always underestimated when you take the highest or lowest
+        # - for images other than Stokes I, where that may apply - pixel.
+        peak *= fudge_max_pix_factor
+    else:
+        peak = data.min()
+
+    rounded_barycenter = int(round(xbar)), int(round(ybar))
+    # basevalue and basepos will be needed for "tweaked moments".
+    try:
+        if not data.mask[rounded_barycenter]:
+            basepos = rounded_barycenter
+        else:
+            basepos = maxpos
+    except IndexError:
+        basepos = maxpos
+    except AttributeError:
+        # If the island is not masked at all, we can safely set basepos to
+        # the rounded barycenter position.
+        basepos = rounded_barycenter
+    basevalue = data[basepos]
+
     # Some problems arise with the sqrt of (working1-working2) when they are
     # equal, this happens with islands that have a thickness of only one pixel
     # in at least one dimension.  Due to rounding errors this difference
     # becomes negative--->math domain error in sqrt.
-    if len(data.nonzero()[0]) == 1:
-        # This is the case when the island (or more likely subisland) has
-        # a size of only one pixel.
-        semiminor = np.sqrt(beamsize / np.pi)
-        semimajor = np.sqrt(beamsize / np.pi)
-    else:
+    if len(data.nonzero()[0]) > 1:
         semimajor_tmp = (working1 + working2) * 2.0 * math.log(2.0)
         semiminor_tmp = (working1 - working2) * 2.0 * math.log(2.0)
-        # ratio will be 0 for data that hasn't been selected according to a
-        # threshold.
-        if ratio != 0:
-            # The corrections below for the semi-major and semi-minor axes are
-            # to compensate for the underestimate of these quantities
-            # due to the cutoff at the threshold.
-            semimajor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
-            semiminor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
-        semimajor = math.sqrt(semimajor_tmp)
-        semiminor = math.sqrt(semiminor_tmp)
-        if semiminor == 0:
+
+        # Theta is not affected by the cut-off at the threshold (see Spreeuw 2010,
+        # page 45).
+        if abs(semimajor_tmp - semiminor_tmp) < 0.01:
+            # short circuit!
+            theta = 0.
+        else:
+            if xxbar != yybar:
+                theta = math.atan(2. * xybar / (xxbar - yybar)) / 2.
+            else:
+                theta = np.sign(xybar) * math.pi / 4.0
+
+            if theta * xybar > 0.:
+                if theta < 0.:
+                    theta += math.pi / 2.0
+                else:
+                    theta -= math.pi / 2.0
+
+        if semiminor_tmp > 0:
+            if np.sign(threshold) == np.sign(basevalue):
+                # Implementation of "tweaked moments", equation 2.67 from
+                # Spreeuw's thesis. In that formula the "base position" was the
+                # maximum pixel position, though. Here that is the rounded
+                # barycenter position, unless it's masked. If it's masked, it
+                # will be the maximum pixel position.
+                deltax, deltay = xbar - basepos[0], ybar - basepos[1]
+
+                epsilon = np.log(2.) * ((np.cos(theta) * deltax +
+                                         np.sin(theta) * deltay) ** 2
+                                        + (np.cos(theta) * deltay -
+                                           np.sin(theta) * deltax) ** 2
+                                        * semiminor_tmp / semimajor_tmp)
+
+                # Set limits for the root finder similar to the bounds for
+                # Gaussian fits.
+                if peak > 0:
+                    low_bound = 0.5 * peak
+                    upp_bound = 1.5 * peak
+                else:
+                    low_bound = 1.5 * peak
+                    upp_bound = 0.5 * peak
+
+                peak = newton_raphson_root_finder(find_true_peak, basevalue,
+                                                  low_bound, upp_bound,
+                                                  threshold, epsilon,
+                                                  semiminor_tmp, basevalue,
+                                                  tol=1e-8, max_iter=100)
+                # The corrections below for the semi-major and semi-minor axes are
+                # to compensate for the underestimate of these quantities
+                # due to the cutoff at the threshold.
+                ratio = threshold / peak
+                semimajor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
+                semiminor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
+
+        elif np.sign(threshold) == np.sign(peak):
             # A semi-minor axis exactly zero gives all kinds of problems.
             # For instance wrt conversion to celestial coordinates.
             # This is a quick fix.
-            semiminor = beamsize / (np.pi * semimajor)
+            ratio = threshold / peak
+            semimajor_tmp /= (1.0 + np.log(ratio) * ratio / (1.0 - ratio))
+            semiminor_tmp = beamsize ** 2 / (np.pi ** 2 * semimajor_tmp)
 
-    if (np.isnan(xbar) or np.isnan(ybar) or
-            np.isnan(semimajor) or np.isnan(semiminor)):
-        raise ValueError("Unable to estimate Gauss shape")
+        semimajor = math.sqrt(semimajor_tmp)
+        semiminor = math.sqrt(semiminor_tmp)
 
-    # Theta is not affected by the cut-off at the threshold (see Spreeuw 2010,
-    # page 45).
-    if abs(semimajor - semiminor) < 0.01:
-        # short circuit!
-        theta = 0.
     else:
-        if xxbar != yybar:
-            theta = math.atan(2. * xybar / (xxbar - yybar)) / 2.
-        else:
-            theta = np.sign(xybar) * math.pi / 4.0
-
-        if theta * xybar > 0.:
-            if theta < 0.:
-                theta += math.pi / 2.0
-            else:
-                theta -= math.pi / 2.0
+        # This is the case when the island (or more likely subisland) has
+        # a size of only one pixel.
+        theta = 0.
+        semiminor = np.sqrt(beamsize / np.pi)
+        semimajor = np.sqrt(beamsize / np.pi)
 
     # NB: a dict should give us a bit more flexibility about arguments;
     # however, all those here are ***REQUIRED***.

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -274,7 +274,7 @@ def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
 
     maxi : float
         Peak pixel value within the island. Units: spectral brightness, 
-        typically Jy/beam. To clarify: source_island[maxposs] == maxi.
+        typically Jy/beam. To clarify: source_island[maxpos] == maxi.
 
     fudge_max_pix_factor : float
         Correction factor for underestimation of the peak by considering the 

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -9,12 +9,12 @@ import scipy.optimize
 from .gaussian import gaussian, jac_gaussian
 from .stats import indep_pixels
 from sourcefinder.deconv import deconv
-from numba import guvectorize, float64, float32, int32
+from numba import guvectorize, float64, float32, int32, njit
 from sourcefinder.stats import newton_raphson_root_finder
 
 FIT_PARAMS = ('peak', 'xbar', 'ybar', 'semimajor', 'semiminor', 'theta')
 
-
+@njit
 def find_true_peak(peak, T, epsilon, msq, maxpix):
     """
     This function represents equation 2.67 from Spreeuw's thesis.

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -10,7 +10,7 @@ from .gaussian import gaussian, jac_gaussian
 from .stats import indep_pixels
 from sourcefinder.deconv import deconv
 from numba import guvectorize, float64, float32, int32, njit
-from sourcefinder.stats import newton_raphson_root_finder
+from sourcefinder.utils import newton_raphson_root_finder
 
 FIT_PARAMS = ('peak', 'xbar', 'ybar', 'semimajor', 'semiminor', 'theta')
 
@@ -148,11 +148,13 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
                     low_bound = 1.5 * peak
                     upp_bound = 0.5 * peak
 
-                peak = newton_raphson_root_finder(find_true_peak, basevalue,
-                                                  low_bound, upp_bound,
-                                                  1e-8, 100,
-                                                  threshold, epsilon,
-                                                  semiminor_tmp, basevalue)
+                # The number of iterations used for the root finder is also
+                # returned, but not used here.
+                peak, _ = newton_raphson_root_finder(find_true_peak, basevalue,
+                                                     low_bound, upp_bound,
+                                                     1e-8, 100,
+                                                     threshold, epsilon,
+                                                     semiminor_tmp, basevalue)
                 # The corrections below for the semi-major and semi-minor axes are
                 # to compensate for the underestimate of these quantities
                 # due to the cutoff at the threshold.

--- a/sourcefinder/fitting.py
+++ b/sourcefinder/fitting.py
@@ -6,6 +6,7 @@ import math
 
 import numpy as np
 import scipy.optimize
+
 from .gaussian import gaussian, jac_gaussian
 from .stats import indep_pixels
 from sourcefinder.deconv import deconv
@@ -35,7 +36,7 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
 
     Parameters
     ----------
-    data : np.ndarray
+    data : np.ma.MaskedArray or np.ndarray
         Actual 2D image data.
     fudge_max_pix_factor : float
         Correct for the underestimation of the peak by taking the maximum
@@ -84,21 +85,6 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
     else:
         peak = data.min()
 
-    rounded_barycenter = int(round(xbar)), int(round(ybar))
-    # basevalue and basepos will be needed for "tweaked moments".
-    try:
-        if not data.mask[rounded_barycenter]:
-            basepos = rounded_barycenter
-        else:
-            basepos = maxpos
-    except IndexError:
-        basepos = maxpos
-    except AttributeError:
-        # If the island is not masked at all, we can safely set basepos to
-        # the rounded barycenter position.
-        basepos = rounded_barycenter
-    basevalue = data[basepos]
-
     # Some problems arise with the sqrt of (working1-working2) when they are
     # equal, this happens with islands that have a thickness of only one pixel
     # in at least one dimension.  Due to rounding errors this difference
@@ -107,8 +93,9 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
         semimajor_tmp = (working1 + working2) * 2.0 * math.log(2.0)
         semiminor_tmp = (working1 - working2) * 2.0 * math.log(2.0)
 
-        # Theta is not affected by the cut-off at the threshold (see Spreeuw 2010,
-        # page 45).
+        # Theta is not affected by the cut-off at the threshold, see Spreeuw's
+        # thesis (2010), page 45, so we can easily compute the position angle
+        # first.
         if abs(semimajor_tmp - semiminor_tmp) < 0.01:
             # short circuit!
             theta = 0.
@@ -125,6 +112,21 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
                     theta -= math.pi / 2.0
 
         if semiminor_tmp > 0:
+            rounded_barycenter = int(round(xbar)), int(round(ybar))
+            # basevalue and basepos will be needed for "tweaked moments".
+            try:
+                if not data.mask[rounded_barycenter]:
+                    basepos = rounded_barycenter
+                else:
+                    basepos = maxpos
+            except IndexError:
+                basepos = maxpos
+            except AttributeError:
+                # If the island is not masked at all, we can safely set basepos to
+                # the rounded barycenter position.
+                basepos = rounded_barycenter
+            basevalue = data[basepos]
+
             if np.sign(threshold) == np.sign(basevalue):
                 # Implementation of "tweaked moments", equation 2.67 from
                 # Spreeuw's thesis. In that formula the "base position" was the
@@ -194,15 +196,15 @@ def moments(data, fudge_max_pix_factor, beamsize, threshold=0):
 
 
 @guvectorize([(float32[:], float32[:], int32[:], int32[:], int32[:], int32,
-               int32, float32, float32, float32, float64, float64, float64[:],
-               float64, float64[:], float64, float64, float32[:, :],
+               int32, float32, float32, int32[:], float32, float64, float64,
+               float64[:], float64, float64[:], float64, float64, float32[:, :],
                float32[:, :], float32[:, :], float32[:, :], float32[:],
                float32[:], float32[:])],
-             ('(n), (n), (m), (n), (n), (), (), (), (), (), (), (), (k), (), ' +
-              ' (m), (), (), (q, r), (q, r), (l, p) -> (l, p), (), (), ()'),
+             ('(n), (n), (m), (n), (n), (), (), (), (), (m), (), (), (), (k),' +
+              '(), (m), (), (), (q, r), (q, r), (l, p) -> (l, p), (), (), ()'),
              nopython=True)
 def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
-                     min_width, no_pixels, threshold, noise, maxi,
+                     min_width, no_pixels, threshold, noise, maxpos, maxi,
                      fudge_max_pix_factor, max_pix_variance_factor, beam,
                      beamsize, correlation_lengths, clean_bias_error,
                      frac_flux_cal_error, Gaussian_islands_map,
@@ -265,9 +267,14 @@ def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
         values at the position of the island's peak pixel value. 
         Units: spectral brightness, typically Jy/beam.
 
+    maxpos : np.ndarray with int32 as dtype and length 2
+        The position of the maximum pixel value within the island, relative
+        to the top-left corner of the rectangular slice encompassing the
+        island. Units: pixels.
+
     maxi : float
         Peak pixel value within the island. Units: spectral brightness, 
-        typically Jy/beam.
+        typically Jy/beam. To clarify: source_island[maxposs] == maxi.
 
     fudge_max_pix_factor : float
         Correction factor for underestimation of the peak by considering the 
@@ -360,14 +367,6 @@ def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
     # The significance of a source detection is determined in this way.
     significance[0] = (source_island / noise_island).max()
 
-    # Are we fitting a -ve or +ve Gaussian?
-    if source_island.mean() >= 0:
-        # The peak is always underestimated when you take the highest pixel.
-        peak = maxi * fudge_max_pix_factor
-    else:
-        peak = source_island.min()
-
-    ratio = threshold / peak
     total = source_island.sum()
     xbar, ybar, xxbar, yybar, xybar = 0, 0, 0, 0, 0
 
@@ -391,16 +390,17 @@ def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
 
     working1 = (xxbar + yybar) / 2.0
     working2 = math.sqrt(((xxbar - yybar) / 2) ** 2 + xybar ** 2)
-    smaj_tmp = (working1 + working2) * 2.0 * math.log(2.0)
-    smin_tmp = (working1 - working2) * 2.0 * math.log(2.0)
-    # ratio will be 0 for data that hasn't been selected according to a
-    # threshold.
-    if ratio != 0:
-        # The corrections below for the semi-major and semi-minor axes are
-        # to compensate for the underestimate of these quantities
-        # due to the cutoff at the threshold.
-        smaj_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
-        smin_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
+
+    if (np.isnan(xbar) or np.isnan(ybar) or np.isnan(working1)
+            or np.isnan(working2)):
+        raise ValueError("Unable to estimate Gauss shape")
+
+    # Are we fitting a -ve or +ve Gaussian?
+    if source_island.mean() >= 0:
+        # The peak is always underestimated when you take the highest pixel.
+        peak = maxi * fudge_max_pix_factor
+    else:
+        peak = source_island.min()
 
     # We need this width to determine Gaussian shape parameters in a meaningful
     # way.
@@ -409,35 +409,106 @@ def moments_enhanced(source_island, noise_island, chunkpos, posx, posy,
         # minimum width of the island, there may still be occasions where
         # working2 can be slightly higher than working1, perhaps due to rounding
         # errors.
+        semimajor_tmp = (working1 + working2) * 2.0 * math.log(2.0)
+        semiminor_tmp = (working1 - working2) * 2.0 * math.log(2.0)
+
+        # Theta is not affected by the cut-off at the threshold, see Spreeuw's
+        # thesis (2010), page 45, so we can easily compute the position angle
+        # first.
+        if abs(semimajor_tmp - semiminor_tmp) < 0.01:
+            # short circuit!
+            theta = 0.
+        else:
+            if xxbar != yybar:
+                theta = math.atan(2. * xybar / (xxbar - yybar)) / 2.
+            else:
+                theta = np.sign(xybar) * math.pi / 4.0
+
+            if theta * xybar > 0.:
+                if theta < 0.:
+                    theta += math.pi / 2.0
+                else:
+                    theta -= math.pi / 2.0
+
+        if semiminor_tmp > 0:
+            # We will be extrapolating from a pixel centred position to the real
+            # position of the Gaussian peak.
+            rounded_barycenter = int(round(xbar)), int(round(ybar))
+            # First we try the pixel centred position closest to the barycenter
+            # position to extrapolate from.
+            if rounded_barycenter[0] in posx and rounded_barycenter[1] in posy:
+                basepos = rounded_barycenter
+                # We need to loop over all possible positions to find the
+                # source_island index corresponding to the rounded barycenter
+                # position.
+                for i in (posx==rounded_barycenter[0]).nonzero()[0]:
+                    if posy[i] == rounded_barycenter[1]:
+                        baseindex = i
+                        basevalue = source_island[baseindex]
+                        break
+            else:
+                # The rounded barycenter position is not in source_island, so we
+                # revert to the maximum pixel position, which is always included.
+                basepos = maxpos[0], maxpos[1]
+                # In this case we do not need to figure out the source_island
+                # index corresponding to the maximum pixel position, because
+                # maxi, the maximum pixel value has already been provided, as an
+                # argument to this function.
+                basevalue = maxi
+
+            if np.sign(threshold) == np.sign(basevalue):
+                # Implementation of "tweaked moments", equation 2.67 from
+                # Spreeuw's thesis. In that formula the "base position" was the
+                # maximum pixel position, though. Here that is the rounded
+                # barycenter position, unless it's masked. If it's masked, it
+                # will be the maximum pixel position.
+                deltax, deltay = xbar - basepos[0], ybar - basepos[1]
+
+                epsilon = np.log(2.) * ((np.cos(theta) * deltax +
+                                         np.sin(theta) * deltay) ** 2
+                                        + (np.cos(theta) * deltay -
+                                           np.sin(theta) * deltax) ** 2
+                                        * semiminor_tmp / semimajor_tmp)
+
+                # Set limits for the root finder similar to the bounds for
+                # Gaussian fits.
+                if peak > 0:
+                    low_bound = 0.5 * peak
+                    upp_bound = 1.5 * peak
+                else:
+                    low_bound = 1.5 * peak
+                    upp_bound = 0.5 * peak
+
+                # The number of iterations used for the root finder is also
+                # returned, but not used here.
+                peak, _ = newton_raphson_root_finder(find_true_peak, basevalue,
+                                                     low_bound, upp_bound,
+                                                     1e-8, 100,
+                                                     threshold, epsilon,
+                                                     semiminor_tmp, basevalue)
+                # The corrections below for the semi-major and semi-minor axes are
+                # to compensate for the underestimate of these quantities
+                # due to the cutoff at the threshold.
+                ratio = threshold / peak
+                semimajor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
+                semiminor_tmp /= (1.0 + math.log(ratio) * ratio / (1.0 - ratio))
+
+        elif np.sign(threshold) == np.sign(peak):
+            # A semi-minor axis exactly zero gives all kinds of problems.
+            # For instance wrt conversion to celestial coordinates.
+            # This is a quick fix.
+            ratio = threshold / peak
+            semimajor_tmp /= (1.0 + np.log(ratio) * ratio / (1.0 - ratio))
+            semiminor_tmp = beamsize ** 2 / (np.pi ** 2 * semimajor_tmp)
+
         try:
-            smaj = math.sqrt(smaj_tmp)
-            smin = math.sqrt(smin_tmp)
+            smaj = math.sqrt(semimajor_tmp)
+            smin = math.sqrt(semiminor_tmp)
             if smin == 0:
                 # A semi-minor axis exactly zero gives all kinds of problems.
                 # For instance wrt conversion to celestial coordinates.
                 # This is a quick fix.
                 smin = beamsize / (np.pi * smaj)
-
-            if (np.isnan(xbar) or np.isnan(ybar) or
-                    np.isnan(smaj) or np.isnan(smin)):
-                raise ValueError("Unable to estimate Gauss shape")
-
-            # Theta is not affected by the cut-off at the threshold (see Spreeuw 2010,
-            # page 45).
-            if abs(smaj - smin) < 0.01:
-                # short circuit!
-                theta = 0.
-            else:
-                if xxbar != yybar:
-                    theta = math.atan(2. * xybar / (xxbar - yybar)) / 2.
-                else:
-                    theta = np.sign(xybar) * math.pi / 4.0
-
-                if theta * xybar > 0.:
-                    if theta < 0.:
-                        theta += math.pi / 2.0
-                    else:
-                        theta -= math.pi / 2.0
 
         # In all cases where we hit a math domain error, we can use the clean
         # beam parameters to derive reasonable estimates for the Gaussian shape

--- a/sourcefinder/image.py
+++ b/sourcefinder/image.py
@@ -43,10 +43,6 @@ def timeit(method):
     return timed
 
 
-def gather(*args):
-    return list(args)
-
-
 logger = logging.getLogger(__name__)
 
 #
@@ -64,6 +60,7 @@ STRUCTURING_ELEMENT = [[1, 1, 1], [1, 1, 1], [1, 1, 1]]  # Island connectiivty
 # Vectorized processing of source islands is much faster, but excludes Gaussian
 # fits, therefore slightly less accurate.
 VECTORIZED = False
+RMS_FILTER = 0.001
 
 
 class ImageData(object):
@@ -478,14 +475,14 @@ class ImageData(object):
 
         if (type(bgmap).__name__ == 'ndarray' or
                 type(bgmap).__name__ == 'MaskedArray'):
-            if bgmap.shape != self.backmap.shape:
+            if bgmap.shape != self.data.shape:
                 raise IndexError("Background map has wrong shape")
             else:
                 self.backmap = bgmap
 
         if (type(noisemap).__name__ == 'ndarray' or
                                        type(noisemap).__name__ == 'MaskedArray'):
-            if noisemap.shape != self.rmsmap.shape:
+            if noisemap.shape != self.data.shape:
                 raise IndexError("Noisemap has wrong shape")
             if noisemap.min() < 0:
                 raise ValueError("RMS noise cannot be negative")
@@ -944,12 +941,17 @@ class ImageData(object):
         # The third filter attempts to exclude those regions of the image
         # which contain no usable data; for example, the parts of the image
         # falling outside the circular region produced by awimager.
-        RMS_FILTER = 0.001
 
-        clipped_data = np.ma.where(
-            (self.data_bgsubbed > analysisthresholdmap) &
-            (self.rmsmap >= (RMS_FILTER * self.grids["rms"].mean())), 1, 0
-        ).filled(fill_value=0)
+        if RMS_FILTER:
+            clipped_data = np.ma.where(
+                (self.data_bgsubbed > analysisthresholdmap) &
+                (self.rmsmap >= (RMS_FILTER * self.grids["rms"].mean())), 1, 0
+            ).filled(fill_value=0)
+        else:
+            clipped_data = np.ma.where(
+                self.data_bgsubbed > analysisthresholdmap, 1, 0
+            ).filled(fill_value=0)
+
         labelled_data, num_labels = ndimage.label(clipped_data,
                                                   STRUCTURING_ELEMENT)
 

--- a/sourcefinder/image.py
+++ b/sourcefinder/image.py
@@ -9,8 +9,6 @@ import logging
 import numpy as np
 from numba import guvectorize, float32, int32
 
-import time
-
 from sourcefinder import extract
 from sourcefinder import stats
 from sourcefinder import utils
@@ -27,21 +25,6 @@ except ImportError:
     from scipy import ndimage
 from numba import guvectorize, float32, int32
 import os
-
-
-def timeit(method):
-    def timed(*args, **kw):
-        ts = time.time()
-        result = method(*args, **kw)
-        te = time.time()
-        if 'log_time' in kw:
-            name = kw.get('log_name', method._name_.upper())
-            kw['log_time'][name] = int((te - ts) * 1000)
-        else:
-            print('{0}  {1:2.2f} ms'.format(method.__name__, (te - ts) * 1000))
-        return result
-    return timed
-
 
 logger = logging.getLogger(__name__)
 
@@ -327,7 +310,6 @@ class ImageData(object):
 
         return {'mean': mean_grid, 'rms': rms_grid, 'indices': centred_inds}
 
-    @timeit
     def _interpolate(self, grid, inds, roundup=False):
 
         """
@@ -422,7 +404,6 @@ class ImageData(object):
     #                                                                         #
     ###########################################################################
 
-    @timeit
     def extract(self, det, anl, noisemap=None, bgmap=None, labelled_data=None,
                 labels=None, deblend_nthresh=0, force_beam=False):
 
@@ -1284,7 +1265,7 @@ class ImageData(object):
                 chunk = slices[label - 1]
 
                 param = extract.ParamSet()
-                param["sig"] = maxis[count] / self.rmsmap.data[tuple(maxposs[count])]
+                param.sig = maxis[count] / self.rmsmap.data[tuple(maxposs[count])]
 
                 param["peak"] = Uncertain(moments_of_sources[count, 0, 0],
                                           moments_of_sources[count, 1, 0])

--- a/sourcefinder/stats.py
+++ b/sourcefinder/stats.py
@@ -178,9 +178,8 @@ def data_clipper_dynamic(flat_data, number_of_non_nan_elements, mean, std):
                 # correct for that.
                 std[0], iterations = (
                     newton_raphson_root_finder(find_true_std, regular_std,
-                                               0, limit,
-                                               regular_std, limit, tol=1e-8,
-                                               max_iter=100))
+                                               0, limit,1e-8,100, regular_std,
+                                               limit))
             else:
                 std[0] = regular_std
 

--- a/sourcefinder/stats.py
+++ b/sourcefinder/stats.py
@@ -24,21 +24,51 @@ def find_true_std(sigma, clipped_std, clip_limit):
 # “This 'derivative' function has been generated using ChatGPT 4.0.
 # Its AI-output has been verified for correctness, accuracy and
 # completeness, adapted where needed, and approved by the author.”
+# @njit
+# def derivative(f, sigma, sigma_meas, D, eps=1e-8):
+#     """
+#     Approximate the derivative of f at sigma using finite differences.
+#     """
+#     delta = eps * sigma
+#     return (f(sigma + delta, sigma_meas, D) - f(sigma, sigma_meas, D)) / delta
+
+
+# “This derivative' function has been generated using
+# ChatGPT 4.0. Its AI-output has been verified for correctness, accuracy and
+# completeness, adapted where needed, and approved by the author.”
 @njit
-def derivative(f, sigma, sigma_meas, D, eps=1e-8):
+def derivative(f, x, eps=1e-8, *args):
     """
     Approximate the derivative of f at sigma using finite differences.
+
+    Parameters
+    ----------
+    f : function
+        The function for which the derivative is to be approximated.
+    x : float
+        The value of the first argument of f at which the derivative is to be
+        approximated.
+    eps : float, default: 1e-8
+        A small value for finite difference approximation.
+    *args : tuple
+        Additional arguments for the function `f`.
+
+    Returns
+    -------
+    float
+        The approximate derivative of f at x.
+
     """
-    delta = eps * sigma
-    return (f(sigma + delta, sigma_meas, D) - f(sigma, sigma_meas, D)) / delta
+    delta = eps * x
+    return (f(x + delta, *args) - f(x, *args)) / delta
 
 
 # “This 'newton_1d_safeguard_sigma' function has been generated using
 # ChatGPT 4.0. Its AI-output has been verified for correctness, accuracy and
 # completeness, adapted where needed, and approved by the author.”
 @njit
-def newton_raphson_root_finder(f, sigma0, sigma_meas, D, min_sigma, max_sigma,
-                               tol=1e-8, max_iter=100):
+def newton_raphson_root_finder(f, sigma0, min_sigma, max_sigma,
+                               tol=1e-8, max_iter=100, *args):
     """
     Solve the transcendental equation for sigma using Newton's method with
     interval safeguards.
@@ -50,18 +80,16 @@ def newton_raphson_root_finder(f, sigma0, sigma_meas, D, min_sigma, max_sigma,
         sigma, sigma_meas, and D.
     sigma0 : float
         Initial guess for the value of sigma.
-    sigma_meas : float
-        Measured sigma value, used in the equation.
-    D : float
-        The parameter D in the transcendental equation.
     min_sigma : float
         Minimum bound for sigma.
     max_sigma : float
         Maximum bound for sigma.
-    tol : float, optional
-        The tolerance for convergence. The default is 1e-8.
-    max_iter : int, optional
-        The maximum number of iterations. The default is 100.
+    tol : float, default: 1e-8
+        The tolerance for convergence.
+    max_iter : int, default: 100
+        The maximum number of iterations.
+    *args : tuple
+        Additional arguments for the function `f`.
 
     Returns
     -------
@@ -90,14 +118,17 @@ def newton_raphson_root_finder(f, sigma0, sigma_meas, D, min_sigma, max_sigma,
     
     sigma = sigma0
     for i in range(max_iter):
-        f_val = f(sigma, sigma_meas, D)
+        # Compute function value
+        f_val = f(sigma, *args)
 
-        df_val = derivative(f, sigma, sigma_meas, D)
+        # Compute numerical derivative
+        f_deriv = derivative(f, sigma, *args, eps=tol)
 
-        if np.abs(df_val) < tol:  # avoid division by zero
+        if np.abs(f_deriv) < tol:  # avoid division by zero
             raise ValueError("Derivative near zero, method fails.")
 
-        delta_sigma = -f_val / df_val
+        # Update sigma using Newton-Raphson method.
+        delta_sigma = -f_val / f_deriv
         sigma += delta_sigma
 
         # Apply safeguard to keep sigma within the interval
@@ -147,7 +178,9 @@ def data_clipper_dynamic(flat_data, number_of_non_nan_elements, mean, std):
                 # correct for that.
                 std[0], iterations = (
                     newton_raphson_root_finder(find_true_std, regular_std,
-                                               regular_std, limit, 0, limit))
+                                               0, limit,
+                                               regular_std, limit, tol=1e-8,
+                                               max_iter=100))
             else:
                 std[0] = regular_std
 

--- a/sourcefinder/stats.py
+++ b/sourcefinder/stats.py
@@ -4,11 +4,10 @@ variances used by the TKP sourcefinder.
 """
 
 import numpy as np
-from numpy.ma import MaskedArray
 from scipy.special import erf
 import numba_scipy
-from scipy.optimize import fsolve
 from numba import njit, guvectorize, int32, float32
+from sourcefinder.utils import newton_raphson_root_finder
 # CODE & NUMBER HANDLING ROUTINES
 
 
@@ -19,129 +18,6 @@ def find_true_std(sigma, clipped_std, clip_limit):
     help2 = np.sqrt(2 * np.pi) * erf(help1)
     return (sigma ** 2 * (help2 - 2 * np.sqrt(2) * help1 *
             np.exp(-help1 ** 2)) - clipped_std ** 2 * help2)
-
-
-# “This 'derivative' function has been generated using ChatGPT 4.0.
-# Its AI-output has been verified for correctness, accuracy and
-# completeness, adapted where needed, and approved by the author.”
-# @njit
-# def derivative(f, sigma, sigma_meas, D, eps=1e-8):
-#     """
-#     Approximate the derivative of f at sigma using finite differences.
-#     """
-#     delta = eps * sigma
-#     return (f(sigma + delta, sigma_meas, D) - f(sigma, sigma_meas, D)) / delta
-
-
-# “This derivative' function has been generated using
-# ChatGPT 4.0. Its AI-output has been verified for correctness, accuracy and
-# completeness, adapted where needed, and approved by the author.”
-@njit
-def derivative(f, x, eps=1e-8, *args):
-    """
-    Approximate the derivative of f at sigma using finite differences.
-
-    Parameters
-    ----------
-    f : function
-        The function for which the derivative is to be approximated.
-    x : float
-        The value of the first argument of f at which the derivative is to be
-        approximated.
-    eps : float, default: 1e-8
-        A small value for finite difference approximation.
-    *args : tuple
-        Additional arguments for the function `f`.
-
-    Returns
-    -------
-    float
-        The approximate derivative of f at x.
-
-    """
-    delta = eps * x
-    return (f(x + delta, *args) - f(x, *args)) / delta
-
-
-# “This 'newton_1d_safeguard_sigma' function has been generated using
-# ChatGPT 4.0. Its AI-output has been verified for correctness, accuracy and
-# completeness, adapted where needed, and approved by the author.”
-@njit
-def newton_raphson_root_finder(f, sigma0, min_sigma, max_sigma,
-                               tol=1e-8, max_iter=100, *args):
-    """
-    Solve the transcendental equation for sigma using Newton's method with
-    interval safeguards.
-
-    Parameters
-    ----------
-    f : function
-        The function to find the root of. It should take three parameters: 
-        sigma, sigma_meas, and D.
-    sigma0 : float
-        Initial guess for the value of sigma.
-    min_sigma : float
-        Minimum bound for sigma.
-    max_sigma : float
-        Maximum bound for sigma.
-    tol : float, default: 1e-8
-        The tolerance for convergence.
-    max_iter : int, default: 100
-        The maximum number of iterations.
-    *args : tuple
-        Additional arguments for the function `f`.
-
-    Returns
-    -------
-    sigma : float
-        The value of sigma that solves the equation, constrained within the 
-        bounds [min_sigma, max_sigma].
-    i : int
-        The number of iterations performed. If convergence was reached, this 
-        is the iteration count. If max iterations were reached, this is equal 
-        to max_iter.
-
-    Raises
-    ------
-    ValueError
-        If the derivative of the function is near zero, causing a division by 
-        zero error.
-        
-    Notes
-    -----
-    The method employs Newton's method for root-finding, with safeguards to 
-    ensure that sigma stays within the bounds [min_sigma, max_sigma]. The 
-    method terminates when the absolute change in sigma is smaller than the 
-    specified tolerance `tol` or when the maximum number of iterations is 
-    reached. If the derivative is too small (near zero), a ValueError is raised.
-    """
-    
-    sigma = sigma0
-    for i in range(max_iter):
-        # Compute function value
-        f_val = f(sigma, *args)
-
-        # Compute numerical derivative
-        f_deriv = derivative(f, sigma, *args, eps=tol)
-
-        if np.abs(f_deriv) < tol:  # avoid division by zero
-            raise ValueError("Derivative near zero, method fails.")
-
-        # Update sigma using Newton-Raphson method.
-        delta_sigma = -f_val / f_deriv
-        sigma += delta_sigma
-
-        # Apply safeguard to keep sigma within the interval
-        # [min_sigma, max_sigma]
-        if sigma < min_sigma:
-            sigma = min_sigma
-        elif sigma > max_sigma:
-            sigma = max_sigma
-
-        if np.abs(delta_sigma) < tol:
-            return sigma, i  # root found, return solution and iterations
-
-    return sigma, max_iter  # max iterations reached, return last estimate
 
 
 @njit

--- a/sourcefinder/utils.py
+++ b/sourcefinder/utils.py
@@ -451,3 +451,86 @@ def two_step_interp(grid, new_xdim, new_ydim):
     interp_per_row(interp_rows.T, x_initial, x_sought, interp_cols.T)
 
     return interp_cols
+
+
+# “This 'newton_raphson_root_finder' function has been generated using
+# ChatGPT 4.0. Its AI-output has been verified for correctness, accuracy and
+# completeness, adapted where needed, and approved by the author.”
+@njit
+def newton_raphson_root_finder(f, sigma0, min_sigma, max_sigma,
+                               tol=1e-8, max_iter=100, *args):
+    """
+    Solve the transcendental equation for sigma using Newton's method with
+    interval safeguards.
+
+    Parameters
+    ----------
+    f : function
+        The function to find the root of. It should take three parameters:
+        sigma, sigma_meas, and D.
+    sigma0 : float
+        Initial guess for the value of sigma.
+    min_sigma : float
+        Minimum bound for sigma.
+    max_sigma : float
+        Maximum bound for sigma.
+    tol : float, default: 1e-8
+        The tolerance for convergence.
+    max_iter : int, default: 100
+        The maximum number of iterations.
+    *args : tuple
+        Additional arguments for the function `f`.
+
+    Returns
+    -------
+    sigma : float
+        The value of sigma that solves the equation, constrained within the
+        bounds [min_sigma, max_sigma].
+    i : int
+        The number of iterations performed. If convergence was reached, this
+        is the iteration count. If max iterations were reached, this is equal
+        to max_iter.
+
+    Raises
+    ------
+    ValueError
+        If the derivative of the function is near zero, causing a division by
+        zero error.
+
+    Notes
+    -----
+    The method employs Newton's method for root-finding, with safeguards to
+    ensure that sigma stays within the bounds [min_sigma, max_sigma]. The
+    method terminates when the absolute change in sigma is smaller than the
+    specified tolerance `tol` or when the maximum number of iterations is
+    reached. If the derivative is too small (near zero), a ValueError is raised.
+    """
+
+    sigma = sigma0
+    for i in range(max_iter):
+        # Evaulate function at current sigma
+        f_val = f(sigma, *args)
+
+        # Compute numerical derivative
+        delta = tol * sigma
+        f_deriv = (f(sigma + delta, *args) - f_val) / delta
+
+        if np.abs(f_deriv) < tol:  # avoid division by zero
+            raise ValueError("Derivative near zero, method fails.")
+
+        # Update sigma using Newton-Raphson method.
+        delta_sigma = -f_val / f_deriv
+        sigma += delta_sigma
+
+        # Apply safeguard to keep sigma within the interval
+        # [min_sigma, max_sigma]
+        if sigma < min_sigma:
+            sigma = min_sigma
+        elif sigma > max_sigma:
+            sigma = max_sigma
+
+        if np.abs(delta_sigma) < tol:
+            return sigma, i  # root found, return solution and iterations
+
+    return sigma, max_iter  # max iterations reached, return last estimate
+

